provisioning/Samples/device/X509Sample/ProvisioningDeviceClientSample.cs

```diff
@@ -0,0 +1,158 @@
+            IAuthenticationMethod auth = new DeviceAuthenticationWithX509Certificate(
+                result.DeviceId,
+                certificate);
+
+            Console.WriteLine($"Testing the provisioned device with IoT Hub...");
+            using DeviceClient iotClient = DeviceClient.Create(result.AssignedHub, auth, _parameters.TransportType);
+
+            Console.WriteLine("Sending a telemetry message...");
+            using var message = new Message(Encoding.UTF8.GetBytes("TestMessage"));
+            await iotClient.SendEventAsync(message);
+
+            Console.WriteLine("Finished.");
+        }
+
+        private ProvisioningTransportHandler GetTransportHandler()
+        {
+            return _parameters.TransportType switch
+            {
+                TransportType.Mqtt => new ProvisioningTransportHandlerMqtt(),
+                TransportType.Mqtt_Tcp_Only => new ProvisioningTransportHandlerMqtt(TransportFallbackType.WebSocketOnly),
+                TransportType.Mqtt_WebSocket_Only => new ProvisioningTransportHandlerMqtt(TransportFallbackType.TcpOnly),
+                TransportType.Amqp => new ProvisioningTransportHandlerAmqp(),
+                TransportType.Amqp_Tcp_Only => new ProvisioningTransportHandlerAmqp(TransportFallbackType.WebSocketOnly),
+                TransportType.Amqp_WebSocket_Only => new ProvisioningTransportHandlerAmqp(TransportFallbackType.TcpOnly),
+                TransportType.Http1 => new ProvisioningTransportHandlerHttp(),
+                _ => throw new NotSupportedException($"Unsupported transport type {_parameters.TransportType}"),
+            };
+        }
+
+        private X509Certificate2 LoadProvisioningCertificate()
+        {
+            ReadCertificatePassword();
+
+            var certificateCollection = new X509Certificate2Collection();
+            certificateCollection.Import(
+                _parameters.GetCertificatePath(),
+                _parameters.CertificatePassword,
+                X509KeyStorageFlags.UserKeySet);
+
+            X509Certificate2 certificate = null;
+
+            foreach (X509Certificate2 element in certificateCollection)
+            {
+                Console.WriteLine($"Found certificate: {element?.Thumbprint} {element?.Subject}; PrivateKey: {element?.HasPrivateKey}");
+                if (certificate == null && element.HasPrivateKey)
+                {
+                    certificate = element;
+                }
+                else
+                {
+                    element.Dispose();
+                }
+            }
+
+            if (certificate == null)
+            {
+                throw new FileNotFoundException($"{_parameters.CertificateName} did not contain any certificate with a private key.");
+            }
+
+            Console.WriteLine($"Using certificate {certificate.Thumbprint} {certificate.Subject}");
+
+            return certificate;
+        }
+
+        private void ReadCertificatePassword()
```

Comment:
I wonder if this is really necessary in the sample code. can't we just throw an error if user hasn't specified a password in the input parameters?

Reply:
If I require it on the command-line, then the password will be visible to anyone watching. This function takes input but doesn't show your password.

The functionality existed before, and I decided to retain it. It seems reasonable to me.

Thoughts?

Reply:
That sounds fair enough. We take connection strings as input parameters all over our sample codes also, one can argue the same logic should be applied elsewhere, but cert passwords are much easier to remember than a SAS token or a key that is much longer ...

Reply:
Good points.